presto-main/src/test/java/com/facebook/presto/operator/BenchmarkPartitionedOutputOperator.java

```diff
@@ -0,0 +1,310 @@
+            for (int i = 0; i < ENTRIES_PER_PAGE; i++) {
+                BIGINT.writeLong(bigintBlockBuilder, i);
+                writeRow(testRows[i], rowBlockBuilder);
+                writeRow(testRows[i], rowBlockBuilder2);
+                writeRow(testRows[i], rowBlockBuilder3);
+            }
+            pageBuilder.declarePositions(ENTRIES_PER_PAGE);
+            return pageBuilder.build();
+        }
+
+        private void writeRow(List<Object> testRow, BlockBuilder rowBlockBuilder)
+        {
+            BlockBuilder singleRowBlockWriter = rowBlockBuilder.beginBlockEntry();
+            for (Object fieldValue : testRow) {
+                if (fieldValue instanceof String) {
+                    VARCHAR.writeSlice(singleRowBlockWriter, utf8Slice((String) fieldValue));
+                }
+                else {
+                    throw new UnsupportedOperationException();
+                }
+            }
+            rowBlockBuilder.closeEntry();
+        }
+
+        // copied & modifed from TestRowBlock
+        private List<Object>[] generateTestRows(List<Type> fieldTypes, int numRows)
+        {
+            List<Object>[] testRows = new List[numRows];
+            for (int i = 0; i < numRows; i++) {
+                List<Object> testRow = new ArrayList<>(fieldTypes.size());
+                for (int j = 0; j < fieldTypes.size(); j++) {
+                    if (fieldTypes.get(j) == VARCHAR) {
+                        byte[] data = new byte[ThreadLocalRandom.current().nextInt(128)];
+                        ThreadLocalRandom.current().nextBytes(data);
+                        testRow.add(new String(data));
+                    }
+                    else {
+                        throw new UnsupportedOperationException();
+                    }
+                }
+                testRows[i] = testRow;
+            }
+            return testRows;
+        }
+
+        private DriverContext createDriverContext()
+        {
+            return TestingTaskContext.builder(EXECUTOR, SCHEDULER, TEST_SESSION)
+                    .setMemoryPoolSize(MAX_MEMORY)
+                    .setSystemMemoryPoolSize(MAX_MEMORY)
+                    .build()
+                    .addPipelineContext(0, true, true)
+                    .addDriverContext();
+        }
+
+        private PartitionedOutputBuffer createPartitionedBuffer(OutputBuffers buffers, DataSize dataSize)
+        {
+            return new PartitionedOutputBuffer(
+                    "task-instance-id",
+                    new StateMachine<>("bufferState", SCHEDULER, OPEN, TERMINAL_BUFFER_STATES),
+                    buffers,
+                    dataSize,
+                    () -> new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext()),
+                    SCHEDULER);
+        }
+    }
+
+    /**
+     * This class doesn't update the memory usage in addInput() while PartitionedOutputOperator does.
+     */
+    static class TestingPartitionedOutputOperator
```

Comment:
I don't think we need this class only for benchmark. It's ok just to have a simple version of the benchmark to show the performance of adding pages to the operator. Then modify the source code of the operator to have or drop the memory updating part.